File: src/agents/shared/confluence.ts
import axios from 'axios';
import { resilientCall } from './resilience';

type ConfluenceApiError = {
  message?: string;
  errors?: Array<{
    message?: string;
    title?: string;
    detail?: string;
  }>;
};

type ConfluencePage = {
  id: string;
  status: string;
  title: string;
  version: {
    number: number;
  };
};

function confluenceClient() {
  const baseURL = process.env.CONFLUENCE_BASE_URL;
  const email = process.env.CONFLUENCE_EMAIL;
  const token = process.env.CONFLUENCE_API_TOKEN;

  if (!baseURL || !email || !token) {
    throw new Error('Missing Confluence environment variables');
  }

  return axios.create({
    baseURL,
    auth: {
      username: email,
      password: token
    },
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  });
}

function getConfluenceErrorMessage(error: unknown) {
  if (!axios.isAxiosError(error)) {
    return error instanceof Error ? error.message : String(error);
  }

  const data = error.response?.data as ConfluenceApiError | undefined;
  const errors = data?.errors
    ?.map((item) => item.message ?? item.title ?? item.detail)
    .filter(Boolean);

  if (errors && errors.length > 0) {
    return errors.join('; ');
  }

  return data?.message ?? `Confluence API request failed with status ${error.response?.status ?? 'unknown'}`;
}

async function confluenceRequest<T>(request: () => Promise<{ data: T }>): Promise<T> {
  return resilientCall('confluence-api', async () => {
    try {
      const response = await request();
      return response.data;
    } catch (error) {
      throw new Error(getConfluenceErrorMessage(error));
    }
  });
}

export async function getConfluencePage(pageId: string): Promise<ConfluencePage> {
  const client = confluenceClient();

  return confluenceRequest<ConfluencePage>(() =>
    client.get(`/wiki/api/v2/pages/${pageId}`, {
      params: {
        'body-format': 'storage'
      }
    })
  );
}

export async function updateConfluencePage(input: {
  pageId: string;
  title: string;
  version: number;
  status?: string;
  html: string;
}) {
  const client = confluenceClient();
  const nextVersion = input.status === 'draft' ? 1 : input.version + 1;

  await confluenceRequest(() =>
    client.put(`/wiki/api/v2/pages/${input.pageId}`, {
      id: input.pageId,
      status: 'current',
      title: input.title,
      body: {
        representation: 'storage',
        value: input.html
      },
      version: {
        number: nextVersion
      }
    })
  );
}

File: python_agents/agents/shared/confluence.py
from __future__ import annotations

import os
from typing import Any

from .http_client import BasicAuthJsonClient


def confluence_configured(require_page: bool = False) -> bool:
    required = ["CONFLUENCE_BASE_URL", "CONFLUENCE_EMAIL", "CONFLUENCE_API_TOKEN"]
    if require_page:
        required.append("CONFLUENCE_PAGE_ID")
    return all(os.environ.get(key) for key in required)


class ConfluenceClient:
    def __init__(self) -> None:
        if not confluence_configured():
            raise RuntimeError("Missing Confluence environment variables")
        self.client = BasicAuthJsonClient(
            os.environ["CONFLUENCE_BASE_URL"].rstrip("/"),
            os.environ["CONFLUENCE_EMAIL"],
            os.environ["CONFLUENCE_API_TOKEN"],
        )

    def get_page(self, page_id: str) -> dict[str, Any]:
        return self.client.request("GET", f"/wiki/api/v2/pages/{page_id}", params={"body-format": "storage"})

    def update_page(self, page_id: str, title: str, version: int, status: str | None, html: str) -> dict[str, Any]:
        next_version = 1 if status == "draft" else version + 1
        return self.client.request("PUT", f"/wiki/api/v2/pages/{page_id}", {
            "id": page_id,
            "status": "current",
            "title": title,
            "body": {
                "representation": "storage",
                "value": html,
            },
            "version": {
                "number": next_version,
            },
        })


def get_confluence_page(page_id: str) -> dict[str, Any]:
    return ConfluenceClient().get_page(page_id)


def update_confluence_page(page_id: str, title: str, version: int, status: str | None, html: str) -> dict[str, Any]:
    return ConfluenceClient().update_page(page_id, title, version, status, html)
